internal/db/schema/postgres.go

```diff
@@ -0,0 +1,352 @@
+					line, col, lineColOK = computeLineFromPos(query, int(pos))
+				}
+			}
+			message := fmt.Sprintf("migration failed")
+			if lineColOK {
+				message = fmt.Sprintf("%s (column %d)", message, col)
+			}
+			if pgErr.Detail != "" {
+				message = fmt.Sprintf("%s, %s", message, pgErr.Detail)
+			}
+			message = fmt.Sprintf("%s, on line %v: %s", message, line, migr)
+			return errors.Wrap(err, op, errors.WithMsg(message))
+		}
+		return errors.Wrap(err, op, errors.WithMsg(fmt.Sprintf("migration failed: %s", migr)))
+	}
+
+	return nil
+}
+
+func computeLineFromPos(s string, pos int) (line uint, col uint, ok bool) {
+	// replace crlf with lf
+	s = strings.Replace(s, "\r\n", "\n", -1)
+	// pg docs: pos uses index 1 for the first character, and positions are measured in characters not bytes
+	runes := []rune(s)
+	if pos > len(runes) {
+		return 0, 0, false
+	}
+	sel := runes[:pos]
+	line = uint(runesCount(sel, newLine) + 1)
+	col = uint(pos - 1 - runesLastIndex(sel, newLine))
+	return line, col, true
+}
+
+const newLine = '\n'
+
+func runesCount(input []rune, target rune) int {
+	var count int
+	for _, r := range input {
+		if r == target {
+			count++
+		}
+	}
+	return count
+}
+
+func runesLastIndex(input []rune, target rune) int {
+	for i := len(input) - 1; i >= 0; i-- {
+		if input[i] == target {
+			return i
+		}
+	}
+	return -1
+}
+
+func (p *postgres) setVersion(ctx context.Context, version int, dirty bool) error {
+	op := errors.Op("schema.setVersion")
+	tx, err := p.conn.BeginTx(ctx, &sql.TxOptions{})
+	if err != nil {
+		return err
+	}
+
+	query := `TRUNCATE ` + pq.QuoteIdentifier(defaultMigrationsTable)
+	if _, err := tx.ExecContext(ctx, query); err != nil {
+		if errRollback := tx.Rollback(); errRollback != nil {
+			err = multierror.Append(err, errRollback)
+		}
+		return errors.Wrap(err, op)
+	}
+
+	// Also re-write the schema version for nil dirty versions to prevent
+	// empty schema version for failed down migration on the first migration
+	// See: https://github.com/golang-migrate/migrate/issues/330
+	if version >= 0 || (version == nilVersion && dirty) {
+		query = `INSERT INTO ` + pq.QuoteIdentifier(defaultMigrationsTable) +
+			` (version, dirty) VALUES ($1, $2)`
+		if _, err := tx.ExecContext(ctx, query, version, dirty); err != nil {
+			if errRollback := tx.Rollback(); errRollback != nil {
+				err = multierror.Append(err, errRollback)
+			}
+			return errors.Wrap(err, op)
+		}
+	}
+
+	if err := tx.Commit(); err != nil {
+		return errors.Wrap(err, op)
+	}
+
+	return nil
+}
+
+func (p *postgres) version(ctx context.Context) (version int, dirty bool, err error) {
+	op := errors.Op("schema.version")
+	query := `SELECT version, dirty FROM ` + pq.QuoteIdentifier(defaultMigrationsTable) + ` LIMIT 1`
+	err = p.conn.QueryRowContext(ctx, query).Scan(&version, &dirty)
+	switch {
+	case err == sql.ErrNoRows:
+		return nilVersion, false, nil
+
+	case err != nil:
+		if e, ok := err.(*pq.Error); ok {
```

Comment:
do we want tests for this code branch?  I don't think, we can get here if `ensureVersionTable(...)` is correct.. but if there's ever a bug in that function...

Reply:
Added a test that exercises that code branch.